src/test/unordered_lru_cache_tests.cpp

```diff
@@ -0,0 +1,402 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <unordered_lru_cache.h>
+
+#include <boost/test/unit_test.hpp>
+
+struct IntHasher {
+    size_t operator()(int v) const noexcept {
+        return std::hash<int>{}(v);
+    }
+};
+
+BOOST_AUTO_TEST_SUITE(unordered_lru_cache_tests)
+
+BOOST_AUTO_TEST_CASE(construction_and_configuration)
+{
+    // Runtime-sized cache with default truncateThreshold
+    unordered_lru_cache<int, int, IntHasher> c1(5);
+    BOOST_CHECK_EQUAL(c1.max_size(), 5U);
+
+    // Custom truncateThreshold
+    unordered_lru_cache<int, int, IntHasher> c2(5, 10);
+    BOOST_CHECK_EQUAL(c2.max_size(), 5U);
+
+    // truncateThreshold less than maxSize
+    unordered_lru_cache<int, int, IntHasher> c3(10, 5);
+    BOOST_CHECK_EQUAL(c3.max_size(), 10U);
+
+    // truncateThreshold equal to maxSize
+    unordered_lru_cache<int, int, IntHasher> c4(5, 5);
+    BOOST_CHECK_EQUAL(c4.max_size(), 5U);
+}
+
+BOOST_AUTO_TEST_CASE(compile_time_maxsize)
+{
+    // Compile-time MaxSize template parameter
+    unordered_lru_cache<int, int, IntHasher, 10> c;
+    BOOST_CHECK_EQUAL(c.max_size(), 10U);
+}
+
+BOOST_AUTO_TEST_CASE(basic_insert_and_get)
+{
+    unordered_lru_cache<int, int, IntHasher> c(10);
+
+    // Insert new key
+    c.insert(1, 10);
+    auto opt = c.get(1);
+    BOOST_CHECK(opt.has_value());
+    BOOST_CHECK_EQUAL(opt.value(), 10);
+
+    // Get non-existent key
+    auto opt2 = c.get(2);
+    BOOST_CHECK(!opt2.has_value());
+
+    // Insert another key
+    c.insert(2, 20);
+    opt2 = c.get(2);
+    BOOST_CHECK(opt2.has_value());
+    BOOST_CHECK_EQUAL(opt2.value(), 20);
+}
+
+BOOST_AUTO_TEST_CASE(emplace_behavior)
+{
+    unordered_lru_cache<int, int, IntHasher> c(10);
+
+    // Emplace new key
+    c.emplace(1, 10);
+    auto opt = c.get(1);
+    BOOST_CHECK(opt.has_value());
+    BOOST_CHECK_EQUAL(opt.value(), 10);
+
+    // Emplace existing key updates value
+    c.emplace(1, 15);
+    opt = c.get(1);
+    BOOST_CHECK(opt.has_value());
+    BOOST_CHECK_EQUAL(opt.value(), 15);
+}
+
+BOOST_AUTO_TEST_CASE(exists_touches_recency)
+{
+    unordered_lru_cache<int, int, IntHasher> c(3, 5);
+
+    // Insert keys 1-5
+    for (int i = 1; i <= 5; ++i) {
+        c.insert(i, i * 10);
+    }
+
+    // Touch key 1 via exists
+    BOOST_CHECK(c.exists(1));
+
+    // Insert key 6, which should trigger truncation
+    // Key 1 should survive because it was touched
+    c.insert(6, 60);
+
+    BOOST_CHECK(c.exists(1)); // Key 1 should still exist
+    BOOST_CHECK(!c.exists(2)); // Key 2 should be evicted
+}
+
+BOOST_AUTO_TEST_CASE(erase_behavior)
+{
+    unordered_lru_cache<int, int, IntHasher> c(10);
+
+    c.insert(1, 10);
+    BOOST_CHECK(c.exists(1));
+
+    c.erase(1);
+    BOOST_CHECK(!c.exists(1));
+    BOOST_CHECK(!c.get(1).has_value());
+
+    // Erasing non-existent key is a no-op
+    c.erase(999);
```

Comment:
should probably insert another element e.g. `(2, 20)` at the start of this test and then check that it wasn't removed neither by `erase(1)` above nor by `erase(999)` here